File: cpp/ch01/E01A_pattern_count.cpp
//
// Created by Matej Hrnčiar on 20/07/2022.
//

#include "E01A_pattern_count.h"

tuple< int, vector<char> > PatternCount(const string &genome, const string &pattern) {
    int count = 0;
    vector<char> indices;

    for (unsigned long i = 0; i < genome.size() - pattern.size(); i++) {
        if (Text(genome, i, pattern.size()) == pattern) {
            count++;
            indices.push_back('^');
        }
        else {
            indices.push_back('-');
        }
    }

    return {count, indices};
}

/*
int main() {
    string _genome, _pattern;

    cout << "Genome: ";
    cin >> _genome;
    cout << "Pattern: ";
    cin >> _pattern;

    auto [count, indices] = PatternCount(_genome, _pattern);

    cout << "Count: " << count << endl;
    cout << _genome << endl;

    for (char c : indices) {
        cout << c;
    }
}
 */

File: py/ch01/E01A_pattern_count.py
# Count the number of occurrences of Pattern in Text
#   Input: A string Text and a string Pattern
#   Output: A number of Pattern occurrences in Text


from utils import text


# Complexity:
# - Cycle: n - k + 1
# - Comparison: k
# O(n - k + 1) * k
def PatternCount(genome, pattern):
    count = 0
    indices = []

    # Cycle through the string and compare each substring with passed pattern
    for i in range(len(genome) - len(pattern) + 1):
        if text(genome, i, len(pattern)) == pattern:
            count += 1
            indices.append("^")
        else:
            indices.append("-")

    return count, indices


if __name__ == "__main__":
    _genome = input("Genome: ").upper()
    _pattern = input("Pattern: ").upper()

    _count, _indices = PatternCount(_genome, _pattern)

    print(_count)
    print(_genome)

    for index in _indices:
        print(index, end="")
